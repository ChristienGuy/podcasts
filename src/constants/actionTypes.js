// Login
export const AUTHENTICATION_LOGIN_ATTEMPT: string = "AUTHENTICATION_LOGIN_ATTEMPT"
export const AUTHENTICATION_LOGIN_FAILURE: string = "AUTHENTICATION_LOGIN_FAILURE"
export const AUTHENTICATION_LOGIN_SUCCESS: string = "AUTHENTICATION_LOGIN_SUCCESS"

// Logout
export const AUTHENTICATION_LOGOUT_FAILURE: string = "AUTHENTICATION_LOGOUT_FAILURE"
export const AUTHENTICATION_LOGOUT_SUCCESS: string = "AUTHENTICATION_LOGOUT_SUCCESS"

// Session
export const AUTHENTICATION_SESSION_CHECK_ATTEMPT: string = "AUTHENTICATION_SESSION_CHECK_ATTEMPT"
export const AUTHENTICATION_SESSION_CHECK_FAILURE: string = "AUTHENTICATION_SESSION_CHECK_FAILURE"
export const AUTHENTICATION_SESSION_CHECK_SUCCESS: string = "AUTHENTICATION_SESSION_CHECK_SUCCESS"


// Podcast
export const PODCAST_SET_PODCASTS: string = "PODCAST_SET_PODCASTS"
export const PODCAST_ADD_PODCAST: string = "PODCAST_ADD_PODCASTS"
export const PODCAST_UPDATE_PODCASTS_ATTEMPT: string = "PODCAST_UPDATE_PODCASTS_ATTEMPT"