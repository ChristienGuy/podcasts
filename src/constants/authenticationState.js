/* 
Tracking states via "enum" rather than boolean flags

Boolean flags lead to potentailly unexpected combinations of states
e.g

isLoggedIn = true
isLoggingIn = false
isCheckingSession = false
isLoadingPodcasts = false
podcastsLoaded = true

This creates 5! number of potential states (1 * 2 * 3 * 4 * 5 = 120)

See: https://christienguy.co.uk/post/react-state-enum/ for full info
*/

// 12 possible combinations of state (3 * 4) is more manageable but still not great
export const AuthenticationState = Object.freeze({
  LOGGING_IN: "LOGGING_IN",
  LOGGED_IN: "LOGGED_IN",
  CHECKING_SESSION: "CHECKING_SESSION",
  NOT_AUTHENTICATED: "NOT_AUTHENTICATED",
})
