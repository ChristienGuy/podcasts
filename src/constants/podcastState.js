export const PodcastState = Object.freeze({
  UPDATING: "UPDATING",
  LOADING: "LOADING",
  LOADED: "LOADED",
  NOT_LOADED: "NOT_LOADED"
})