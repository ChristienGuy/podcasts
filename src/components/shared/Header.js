import React, { Fragment } from "react";
import styled from "styled-components";

import { Link } from "react-router-dom";

import { AuthenticationState } from "../../constants/authenticationState";
import { PodcastState } from "../../constants/podcastState";
import PropTypes from "prop-types";

const Wrapper = styled.div`
  display: flex;
  align-items: center;
  justify-content: flex-end;
  height: 60px;
  background-color: #222;
  padding-left: 20px;
  padding-right: 20px;

  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
`;

const WhiteLink = styled(Link)`
  color: #fff;
  margin-right: 20px;
`;

export const Header = ({ logout, authentication, podcastState }) => {
  return (
    <Wrapper>
      { podcastState === PodcastState.UPDATING && <span>Updating Podcasts</span> }
      <WhiteLink to="/">Home</WhiteLink>
      {authentication.authenticationState === AuthenticationState.LOGGED_IN ? (
        <Fragment>
          <WhiteLink to="/podcast/add">Add Podcast</WhiteLink>
          <LogoutButton onLogout={logout} />
        </Fragment>
      ) : (
        <WhiteLink to="/login">Login</WhiteLink>
      )}
    </Wrapper>
  );
};

export const LogoutButton = ({onLogout}) => (
  <button onClick={onLogout}>Logout</button>
)
LogoutButton.displayName = "LogoutButton";

Header.propTypes = {
  logout: PropTypes.func.isRequired,
  authentication: PropTypes.object,
  podcastState: PropTypes.string
}

export default Header;
