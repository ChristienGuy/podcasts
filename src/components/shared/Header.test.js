import React from "react";
import { Header } from "./Header";
import { shallow, mount } from "enzyme";

import { AuthenticationState } from "../../constants/authenticationState";
import { PodcastState } from "../../constants/podcastState";

const setup = addedProps => {
  const props = {
    logout: jest.fn(),
    authentication: {},
    podcastState: "",
    ...addedProps
  };

  const wrapper = shallow(<Header {...props} />);

  return {
    props,
    wrapper
  };
};

it("renders without crashing", () => {
  const { wrapper } = setup();
  expect(wrapper).toMatchSnapshot();
});

it("Should render logout button when the user is logged in", () => {
  const { wrapper, props } = setup({
    authentication: { authenticationState: AuthenticationState.LOGGED_IN }
  });

  expect(wrapper.find('LogoutButton')).toBeTruthy()
  expect(wrapper).toMatchSnapshot();
});

it("should call logout when logout is clicked", () => {
  const { wrapper, props } = setup({
    authentication: { authenticationState: AuthenticationState.LOGGED_IN }
  });
  const button = wrapper.find("LogoutButton");

  button.props().onLogout();
  expect(props.logout.mock.calls.length).toBe(1);
});

it("should show 'updating podcasts' when podcastState === UPDATING", () => {
  const { wrapper, props } = setup({
    authentication: { podcastState: PodcastState.UPDATING }
  });
  
  expect(wrapper).toMatchSnapshot()
});
