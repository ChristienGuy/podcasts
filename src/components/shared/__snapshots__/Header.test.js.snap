// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render logout button when the user is logged in 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    Home
  </Styled(Link)>
  <React.Fragment>
    <Styled(Link)
      to="/podcast/add"
    >
      Add Podcast
    </Styled(Link)>
    <LogoutButton
      onLogout={[Function]}
    />
  </React.Fragment>
</styled.div>
`;

exports[`renders without crashing 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    Home
  </Styled(Link)>
  <Styled(Link)
    to="/login"
  >
    Login
  </Styled(Link)>
</styled.div>
`;

exports[`should show 'updating podcasts' when podcastState === UPDATING 1`] = `
<styled.div>
  <Styled(Link)
    to="/"
  >
    Home
  </Styled(Link)>
  <Styled(Link)
    to="/login"
  >
    Login
  </Styled(Link)>
</styled.div>
`;
